packages/liveblocks-core/src/lib/position.ts

```diff
@@ -1,133 +1,340 @@
+ * representation uses base 96.
+ *
+ *   0 ≃ ' '  (lowest digit)
+ *   1 ≃ '!'
+ *   2 ≃ '"'
+ *   ...
+ *   9 ≃ '~'  (highest digit)
+ *
+ * So think:
+ *   '!'    ≃ 0.1
+ *   '"'    ≃ 0.2
+ *   '!"~'  ≃ 0.129
+ *
+ * Three rules:
+ * - All "characters" in the string should be valid digits (from the above
+ *   alphabet)
+ * - The value 0.0 is not a valid Pos value
+ * - A Pos cannot have trailing "zeroes"
+ *
+ * This representation has the following benefits:
+ *
+ * 1. It's always possible to get a number that lies before, after, or between
+ *    two arbitrary Pos values.
+ * 2. Pos values can be compared using normal string comparison.
+ *
+ * Some examples:
+ * - '!'  < '"'   (like how .1  < .2)
+ * - '!'  < '~'   (like how .1  < .9)
+ * - '!!' < '!~'  (like how .11 < .19)
+ * - '~!' < '~~'  (like how .91 < .99)
+ * - '~'  < '~!'  (like how .9  < .91)
+ * - '!!' < '!O'  (like how .1  < .5)
+ * - '!O' < '!~'  (like how .5  < .9)
+ *
+ */
+import type { Brand } from "./utils";
 
-export function makePosition(before?: string, after?: string): string {
-  // Between
-  if (before !== undefined && after !== undefined) {
-    return pos(makePositionFromCodes(posCodes(before), posCodes(after)));
-  }
+/**
+ * A valid/verified "position" string. These values are used as "parentKey"s by
+ * LiveList children, and define their relative ordering.
+ */
+export type Pos = Brand<string, "Pos">;
 
-  // Insert at the end
-  else if (before !== undefined) {
-    return getNextPosition(before);
-  }
+const minCode = 32;
+const maxCode = 126;
+
+const NUM_DIGITS = maxCode - minCode + 1;
+
+const ZERO: string = nthDigit(0); // " "
+
+/**
+ * The "first" canonical position.
+ * In an equivalent decimal number system, think of this as the value .1.
+ */
+const ONE: Pos = nthDigit(1); // "!"
 
-  // Insert at the start
-  else if (after !== undefined) {
-    return getPreviousPosition(after);
+const ZERO_NINE = (ZERO + nthDigit(-1)) as Pos;
+
+/**
+ * Returns the Pos value for the nth digit in the alphabet.
+ * Value must be between 0 and 94.
+ *
+ * Just used to generate some static data, and for usage in test cases.
+ */
+function nthDigit(n: 0): string; // "0" is a legal _digit_, but not a legal Pos value
+function nthDigit(n: number): Pos;
+function nthDigit(n: number): Pos {
+  const code = minCode + (n < 0 ? NUM_DIGITS + n : n);
+  if (code < minCode || code > maxCode) {
+    throw new Error(`Invalid n value: ${n}`);
   }
+  return String.fromCharCode(code) as Pos;
+}
 
-  // No children
-  return pos([min + 1]);
+/**
+ * Given two positions, returns the position value that lies in the middle.
+ * When given only a high bound, computes the canonical position "before" it.
+ * When given only a low bound, computes the canonical position "after" it.
+ * When given no bounds at all, returns the "first" canonical position.
+ */
+function makePosition(x?: Pos, y?: Pos): Pos {
+  return x !== undefined && y !== undefined
```

Comment:
I personally don't mind nested ternaries, but I was once given the advice to never nest more than one because they become very hard to read.

Reply:
Completely agree, there is no reason to use this here. Fixed in a6b93bf3723128aa8865b375e7dc1a609cd13ec1.